src/main/java/beam/physsim/jdeqsim/cacc/roadCapacityAdjustmentFunctions/Hao2018CaccRoadCapacityAdjustmentFunction.java

```diff
@@ -72,94 +76,101 @@ public Hao2018CaccRoadCapacityAdjustmentFunction(BeamConfig beamConfig,int itera
-        double initialCapacity=link.getCapacity();
-        return initialCapacity>=caccMinRoadCapacity && link.getFreespeed()>=caccMinSpeedMetersPerSec;
+    public boolean isCACCCategoryRoad(Link link) {
+        double initialCapacity = link.getCapacity();
+        return initialCapacity >= caccMinRoadCapacity && link.getFreespeed() >= caccMinSpeedMetersPerSec;
     }
 
-    public double getCapacityWithCACCPerSecond(Link link, double fractionCACCOnRoad,double simTime){
-        double initialCapacity=link.getCapacity();
-        double updatedCapacity=initialCapacity;
+    public double getCapacityWithCACCPerSecond(Link link, double fractionCACCOnRoad, double simTime) {
+        double initialCapacity = link.getCapacity();
+        double updatedCapacity = initialCapacity;
 
         if (isCACCCategoryRoad(link)) {
             caccCategoryRoadsTravelled++;
-            updatedCapacity= calculateCapacity(fractionCACCOnRoad, initialCapacity);
+            updatedCapacity = calculateCapacity(fractionCACCOnRoad, initialCapacity);
 
 
-
-            if (fractionCACCOnRoad==1){
+            if (fractionCACCOnRoad == 1) {
                 numberOfTimesOnlyCACCTravellingOnCACCEnabledRoads++;
             }
 
-            if (fractionCACCOnRoad==0){
+            if (fractionCACCOnRoad == 0) {
                 numberOfTimesOnlyNonCACCTravellingOnCACCEnabledRoads++;
             }
 
-            if (fractionCACCOnRoad>0 && fractionCACCOnRoad<=1.0){
+            if (fractionCACCOnRoad > 0 && fractionCACCOnRoad <= 1.0) {
                 numberOfMixedVehicleTypeEncountersOnCACCCategoryRoads++;
-                capacityIncreaseSum+=updatedCapacity-initialCapacity;
-                percentageCapacityIncreaseSum+=(updatedCapacity/initialCapacity-1.0);
+                capacityIncreaseSum += updatedCapacity - initialCapacity;
+                percentageCapacityIncreaseSum += (updatedCapacity / initialCapacity - 1.0);
             }
 
 
-            if (updatedCapacity<initialCapacity){
-                log.error("updatedCapacity (" + updatedCapacity +") is lower than initialCapacity (" + initialCapacity + ").");
+            if (updatedCapacity < initialCapacity) {
+                log.error("updatedCapacity (" + updatedCapacity + ") is lower than initialCapacity (" + initialCapacity + ").");
             }
 
-            String dataLine = link.getId().toString() + "," + fractionCACCOnRoad + "," + initialCapacity + "," + updatedCapacity;
-            capacityStatsCollector.append(dataLine).append("\n");
-
-            double capacityIncreaseForCACCEnabledRoads = (updatedCapacity/initialCapacity)-1.0;
-            caccCapacityIncrease.put(fractionCACCOnRoad * 100.0,capacityIncreaseForCACCEnabledRoads * 100.0);
-            caccLinkCapacityIncrease.put(link.getId().toString(),capacityIncreaseForCACCEnabledRoads * 100.0);
+            double finalUpdatedCapacity = updatedCapacity;
+            csvWriter.ifPresent(writer -> {
```

Comment:
Why should I instantiate it here? It is like while loop and this method is called multiple times

Reply:
this method is bad smell. It is called `getCapacityWithCACCPerSecond` but it actually `writes`. The writer is close by the method `reset` which is a bad smell as well. The resource will be kept open in case the caller does not close it. If I understood we have a public method doing called `get` and generating a side effect. This should be refactored somehow

Reply:
Please, take a look. I made sure that `clear` is called in any case

Reply:
I saw the AutoCloseable comment first, and then this one, and I agree with Carlos. This is primed for a memory leak...we shouldn't leave the door open even a little for those. Moreso, we should maybe be explicit of the side effecting that occurs and make it a passed in function?

Reply:
The main reason why I don't create `CsvMapWriter` every time in `getCapacityWithCACCPerSecond` because it is called in every Leave/Enter road message during JDEQSim(PhysSim). Stacktrace:
```
java.lang.OutOfMemoryError: Requested array size exceeds VM limit
        at java.util.Arrays.copyOf(Arrays.java:3332)
        at java.lang.AbstractStringBuilder.ensureCapacityInternal(AbstractStringBuilder.java:124)
        at java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:448)
        at java.lang.StringBuffer.append(StringBuffer.java:270)
        at beam.physsim.jdeqsim.cacc.roadCapacityAdjustmentFunctions.Hao2018CaccRoadCapacityAdjustmentFunction.getCapacityWithCACCPerSecond(Hao2018CaccRoadCapacityAdjustmentFunction.java:112)
        at beam.physsim.jdeqsim.cacc.sim.Road.getInverseCapacity(Road.java:120)
        at beam.physsim.jdeqsim.cacc.sim.Road.leaveRoad(Road.java:197)
        at org.matsim.core.mobsim.jdeqsim.LeaveRoadMessage.handleMessage(LeaveRoadMessage.java:36)
        at org.matsim.core.mobsim.jdeqsim.Scheduler.startSimulation(Scheduler.java:63)
```
So this could slow down that part of simulation a lot

Reply:
Fair enough. Perf trumps better code often.